### ISSUE TITLE
Dir[] vs "../../"

### ISSUE THREAD

**Issue by jmettraux** [jruby/jruby#8426]
**Environment Information**
  
* jruby 9.4.9.0 (3.1.4) 2024-11-04 547c6b150e OpenJDK 64-Bit Server VM 21.0.4+7-1 on 21.0.4+7-1 +jit [x86_64-OpenBSD]
* OpenBSD armata 7.6 GENERIC.MP#338 amd64

and

*  Windows Server 2019 10.0 amd64 JVM 11.0.23+9-LTS Azul Systems, Inc.

**Expected Behavior**

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.
* Provide an executable Ruby script or a link to an example repository.

```ruby
pp Dir[File.join('../../t', '**', 'site.css')]

# ruby 3.3.5 (2024-09-03 revision ef084cc8f4) [x86_64-openbsd]
#
# [ "../../t/sg6/public/sheets/site.css",
#   "../../t/sg6/war_20240407_1734/sheets/site.css" ]
```

**Actual Behavior**

```ruby

pp Dir[File.join('../../t', '**', 'site.css')]

# jruby 9.4.9.0 (3.1.4) 2024-11-04 547c6b150e OpenJDK 64-Bit Server
#   VM 21.0.4+7-1 on 21.0.4+7-1 +jit [x86_64-OpenBSD]
#
# [ "./aux/t/sg6/public/sheets/site.css",
#   "./aux/t/sg6/war_20240407_1734/sheets/site.css"]
```

**Comment by headius** [comment]
Looks like it's taking the `../..` and making that the current dir, so then the results it finds there resolve as though they are relative to that directory.

It turns out to be even weirder than that though:

```
$ jruby -e 'pp Dir[File.join("../../bin", "**")]'
["../../bin/addbin.sh", "../../bin/md5sum", "../../bin/pickjdk.bash", "../../bin/sha1sum", "../../bin/sha256sum", "../../bin/sha512sum"]
$ jruby -e 'pp Dir[File.join("../../bin", "**", "addbin.sh")]'
["./ius/bin/addbin.sh"]
```

The first example, without a specific filename provided after the `"**"` element to the path, resolves as you expect. When we add a filename, a la `"../../**/addbin.sh"` we get the unusual CWD logic... but it's actually *cropping* part of the path. The full path should be relative to my home dir, `/Users/headius` but instead we get a path of `./ius/bin/addbin.sh`.

Interestingly, "./ius/" has the same number of characters as "../../". We may be trying to inject the relative path back in but using part of the full path to do it.

In your original example, is the "aux" dir actually named "aux"?

**Comment by headius** [comment]
Specifically in my case I believe this line of code:

https://github.com/jruby/jruby/blob/47ad4a5724bc0634968d4cb3d7ad13654c445a30/core/src/main/java/org/jruby/util/Dir.java#L791

... takes the incoming full path "/Users/headius/bin/addbin.sh" and truncates it to the length of the relative path "../../bin/addbin.sh", adding a prefix "./" and resulting in the bogus path "./ius/bin/addbin.sh".

**Comment by jmettraux** [comment]
> In your original example, is the "aux" dir actually named "aux"?

No, not at all. Your username is `headius` hence `"./ius/bin/addbin.sh"`. My username is `jmettraux` hence `"./aux/t/sg6/public/sheets/site.css"`.

**Comment by headius** [comment]
@jmettraux Of course!

**Comment by headius** [comment]
Looks like @edipofederle may not be available so we should try to patch this ourselves.

**Comment by danini-the-panini** [comment]
I think #6681 was not the correct solve for cased case-insensitive filesystem, and that there is a deeper issue afoot with the underlying globbing logic. For example:

MRI:

```
irb(main):021> Dir['lorem*/foo.txt']
=> ["LoremIpsum/Foo.txt"]
```

JRuby:

```
irb(main):008:0> Dir['lorem*/foo.txt']
=> []
```

**Comment by headius** [comment]
@danini-the-panini Perhaps we should revert it and try a new attempt?

**Comment by danini-the-panini** [comment]
I'll have to go do a deep dive of dir.c and see how it's done in MRI

**Comment by headius** [comment]
Godspeed! It's tangled logic.

Depending on your priority we could try another small one-off fix, but if we can do it *correctly* that would obviously be better.

**Comment by danini-the-panini** [comment]
Interestingly this case sensitive pattern issue does not happen on Windows

Windows:
```
irb(main):012:0> Dir['lorem/foo.txt']
=> ["Lorem/foo.txt"]
irb(main):013:0> Dir['lore*/foo.txt']
=> ["Lorem/foo.txt"]
```

macOS:
```
irb(main):007:0> Dir['lorem/foo.txt']
=> ["Lorem/foo.txt"]
irb(main):008:0> Dir['lore*/foo.txt']
=> []
```

**Comment by headius** [comment]
They are different file systems, but both are merely case preserving so I'm not sure why there would be such a difference. We do use some JDK apis to implement this so perhaps it's a difference at that level?

**Comment by enebo** [comment]
There is a lot of weird logic with case on windows:

```java
public final static boolean DOSISH = Platform.IS_WINDOWS;
                            if (DOSISH && (pathname && isdirsep(pbytes[p]) && isdirsep(sbytes[s]))) {
                            } else {
                                if (!nocase) {
                                    break; // branch to failed
                                }

                                // TODO: Use JOni case folding
                                if (Character.toLowerCase(codePoint(enc, pbytes, p, pend)) !=
                                        Character.toLowerCase(codePoint(enc, sbytes, s, send))) {
                                    break; // branch to failed
                                }

                            }
```

I have wondered if we need to worry about the same stuff when this was ported in the early 00s.  Modern NTFS systems may support case sensitivity better?

**Comment by danini-the-panini** [comment]
Ok it looks like JRuby is enabling FNM_CASEFOLD by default on Windows. If I explicitly pass the FNM_CASEFOLD flag on macOS it solves some problems:

```
irb(main):001:0> Dir.glob('lore*/foo.txt', File::FNM_CASEFOLD)
=> ["Lorem/foo.txt"]
```

However FNM_CASEFOLD behaviour seems to be different again on case-sensitive filesystems

JRuby (APFS Case-sensitive)
```
irb(main):006:0> Dir.glob('lorem/**/*.txt')
=> []
irb(main):007:0> Dir.glob('lorem/**/*.txt', File::FNM_CASEFOLD)
=> ["Lorem/Foo.txt", "Lorem/Ipsum/Foo.txt", "Lorem/Ipsum/foo.txt", "Lorem/foo.txt"]
irb(main):008:0> Dir.glob('Lorem/**/*.txt')
=> ["Lorem/Foo.txt", "Lorem/Ipsum/Foo.txt", "Lorem/Ipsum/foo.txt", "Lorem/foo.txt"]
```

MRI (APFS Case-sensitive)
```
irb(main):001> Dir.glob('lorem/**/*.txt')
=> []
irb(main):002> Dir.glob('lorem/**/*.txt', File::FNM_CASEFOLD)
=> []
irb(main):003> Dir.glob('Lorem/**/*.txt')
=> ["Lorem/Foo.txt", "Lorem/Ipsum/Foo.txt", "Lorem/Ipsum/foo.txt", "Lorem/foo.txt"]
```

So far digging into dir.c has revealed that it uses getattrlist on macOS to determine case-sensitivity, then dynamically enables FNM_CASEFOLD inside glob_helper.

@enebo Microsoft added a per-directory case-sensitivity attribute to NTFS in 2018, but I don't think MRI bothers with it. Seems like a niche feature for WSL interop

**Comment by headius** [comment]
Interesting findings! We can probably do something similar to getattrlist using the JDK file system API to query for metadata. I'm betting we just never even thought to enable file system case folding on MacOS.

**Comment by danini-the-panini** [comment]
I can't seem to find a way to read that specific attribute via the JDK. From what I can gather, openjdk just assumes case insensitivity on MacOS, and doesn't expose a way to query it per-volume. The only other way I can see us getting around this is calling `volumeSupportsCaseSensitiveNames` via JNA.

Perhaps we should just default to FNM_CASEFOLD on macOS for now until we come up with a more accurate approach.

I've also discovered that explicitly passing in FNM_CASEFOLD is supposed to have no effect on Dir.glob (it's only meant for fnmatch) 🤦🏻‍♀️ 

And... I found another bug: FNM_DOTMATCH matches `.` in MRI but not JRuby:

MRI:
```
irb(main):004> Dir.glob('lorem/*', File::FNM_DOTMATCH)
=> 
["Lorem/.",
 "Lorem/.DS_Store",
 "Lorem/Dolor",
 "Lorem/Ipsum",
 "Lorem/a",
 "Lorem/ab",
 "Lorem/foo.txt"]
```

JRuby:
```
irb(main):002:0> Dir.glob('lorem/*', File::FNM_DOTMATCH)
["Lorem/.DS_Store",
 "Lorem/Dolor",
 "Lorem/Ipsum",
 "Lorem/a",
 "Lorem/ab",
 "Lorem/foo.txt"]
```

**Comment by edipofederle** [comment]
Sorry @headius I Ihave a newborn at home now :D so I missed the messages. I'll read the previous messages.

**Comment by headius** [comment]
Looks like we need more investigation and work here so this is punted to 9.4.11.0.

@danini-the-panini I also could not find any way to get the case-folding attribute for a filesystem from any JDK APIs. A workaround I saw posted several places is to take the filename, change the case, and ask the API if it's the same file. There's internal logic that treats MacOS filesystems as case-preserving. That path might be our simplest option for now.

Do you have some time to help me put together a PR? Feel free to contact me outside GitHub (like on our Matrix channel) and we'll discuss what needs to be done.

**Comment by danini-the-panini** [comment]
Yeah, and we can't use the "make and rename a file" trick because it would fail if a program has no write access to that specific volume anyway 😑

It's tricky to chat on matrix since our time zones are so wildly different. We might be able to catch each other in the evening. Personally, I've found it easier to communicate async here on Github.

**Comment by headius** [comment]
@danini-the-panini Whatever works for you!
